File: options_backtester/execution/cost_model.py
"""Transaction cost models for options and stocks."""

from __future__ import annotations

from abc import ABC, abstractmethod


class TransactionCostModel(ABC):
    """Base class for all transaction cost models."""

    @abstractmethod
    def option_cost(self, price: float, quantity: int, shares_per_contract: int) -> float:
        """Return total commission for an options trade."""
        ...

    @abstractmethod
    def stock_cost(self, price: float, quantity: float) -> float:
        """Return total commission for a stock trade."""
        ...


class NoCosts(TransactionCostModel):
    """Zero transaction costs — matches original behavior."""

    def option_cost(self, price: float, quantity: int, shares_per_contract: int) -> float:
        return 0.0

    def stock_cost(self, price: float, quantity: float) -> float:
        return 0.0


class PerContractCommission(TransactionCostModel):
    """Fixed per-contract commission (e.g., $0.65/contract for IBKR)."""

    def __init__(self, rate: float = 0.65, stock_rate: float = 0.005) -> None:
        self.rate = rate
        self.stock_rate = stock_rate  # per-share

    def option_cost(self, price: float, quantity: int, shares_per_contract: int) -> float:
        return self.rate * abs(quantity)

    def stock_cost(self, price: float, quantity: float) -> float:
        return self.stock_rate * abs(quantity)


class TieredCommission(TransactionCostModel):
    """Tiered commission schedule with volume discounts.

    Tiers are (max_contracts, rate) pairs sorted by max_contracts ascending.
    Contracts beyond the last tier use the last tier's rate.
    """

    def __init__(self, tiers: list[tuple[int, float]] | None = None,
                 stock_rate: float = 0.005) -> None:
        # Default: IBKR-style tiers
        self.tiers = tiers or [
            (10_000, 0.65),
            (50_000, 0.50),
            (100_000, 0.25),
        ]
        self.stock_rate = stock_rate

    def option_cost(self, price: float, quantity: int, shares_per_contract: int) -> float:
        qty = abs(quantity)
        total = 0.0
        remaining = qty
        prev_bound = 0
        for max_qty, rate in self.tiers:
            tier_qty = min(remaining, max_qty - prev_bound)
            if tier_qty <= 0:
                prev_bound = max_qty
                continue
            total += tier_qty * rate
            remaining -= tier_qty
            prev_bound = max_qty
            if remaining <= 0:
                break
        if remaining > 0:
            total += remaining * self.tiers[-1][1]
        return total

    def stock_cost(self, price: float, quantity: float) -> float:
        return self.stock_rate * abs(quantity)


class SpreadSlippage(TransactionCostModel):
    """Model slippage as a fraction of the bid-ask spread.

    Example: SpreadSlippage(pct=0.5) means you pay half the spread on top
    of the execution price.
    """

    def __init__(self, pct: float = 0.5) -> None:
        assert 0.0 <= pct <= 1.0
        self.pct = pct

    def option_cost(self, price: float, quantity: int, shares_per_contract: int) -> float:
        # Slippage is modeled separately via fill_model; this returns 0
        # so it can be composed with a commission model.
        return 0.0

    def stock_cost(self, price: float, quantity: float) -> float:
        return 0.0

    def slippage(self, bid: float, ask: float, quantity: int, shares_per_contract: int) -> float:
        """Compute dollar slippage from the spread."""
        spread = abs(ask - bid)
        return self.pct * spread * abs(quantity) * shares_per_contract

File: rust/ob_core/src/cost_model.rs
//! Transaction cost models for options and stocks.
//!
//! Mirrors Python's `options_portfolio_backtester.execution.cost_model`.

#[derive(Debug, Clone, Default)]
pub enum CostModel {
    /// Zero transaction costs.
    #[default]
    NoCosts,
    /// Fixed per-contract commission (e.g., $0.65/contract for IBKR).
    PerContract { rate: f64, stock_rate: f64 },
    /// Tiered commission schedule with volume discounts.
    /// Tiers are (max_contracts, rate) pairs sorted by max_contracts ascending.
    Tiered { tiers: Vec<(i64, f64)>, stock_rate: f64 },
}

impl CostModel {
    /// Compute option trade commission.
    pub fn option_cost(&self, _price: f64, quantity: f64, _spc: i64) -> f64 {
        let qty = quantity.abs();
        match self {
            CostModel::NoCosts => 0.0,
            CostModel::PerContract { rate, .. } => rate * qty,
            CostModel::Tiered { tiers, .. } => {
                let mut total = 0.0;
                let mut remaining = qty;
                let mut prev_bound: i64 = 0;
                for &(max_qty, rate) in tiers {
                    let tier_qty = remaining.min((max_qty - prev_bound) as f64);
                    if tier_qty <= 0.0 {
                        prev_bound = max_qty;
                        continue;
                    }
                    total += tier_qty * rate;
                    remaining -= tier_qty;
                    prev_bound = max_qty;
                    if remaining <= 0.0 {
                        break;
                    }
                }
                if remaining > 0.0 {
                    if let Some(&(_, last_rate)) = tiers.last() {
                        total += remaining * last_rate;
                    }
                }
                total
            }
        }
    }

    /// Compute stock trade commission.
    pub fn stock_cost(&self, _price: f64, quantity: f64) -> f64 {
        let qty = quantity.abs();
        match self {
            CostModel::NoCosts => 0.0,
            CostModel::PerContract { stock_rate, .. } => stock_rate * qty,
            CostModel::Tiered { stock_rate, .. } => stock_rate * qty,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_costs() {
        let m = CostModel::NoCosts;
        assert_eq!(m.option_cost(10.0, 5.0, 100), 0.0);
        assert_eq!(m.stock_cost(150.0, 100.0), 0.0);
    }

    #[test]
    fn per_contract() {
        let m = CostModel::PerContract { rate: 0.65, stock_rate: 0.005 };
        assert!((m.option_cost(10.0, 10.0, 100) - 6.5).abs() < 1e-10);
        assert!((m.option_cost(10.0, -10.0, 100) - 6.5).abs() < 1e-10);
        assert!((m.stock_cost(150.0, 100.0) - 0.5).abs() < 1e-10);
    }

    #[test]
    fn tiered_within_first_tier() {
        let m = CostModel::Tiered {
            tiers: vec![(10_000, 0.65), (50_000, 0.50), (100_000, 0.25)],
            stock_rate: 0.005,
        };
        // 100 contracts, all in first tier
        assert!((m.option_cost(10.0, 100.0, 100) - 65.0).abs() < 1e-10);
    }

    #[test]
    fn tiered_spanning_tiers() {
        let m = CostModel::Tiered {
            tiers: vec![(10_000, 0.65), (50_000, 0.50), (100_000, 0.25)],
            stock_rate: 0.005,
        };
        // 15000 contracts: 10000 * 0.65 + 5000 * 0.50
        let expected = 10_000.0 * 0.65 + 5_000.0 * 0.50;
        assert!((m.option_cost(10.0, 15_000.0, 100) - expected).abs() < 1e-10);
    }

    #[test]
    fn tiered_beyond_all() {
        let m = CostModel::Tiered {
            tiers: vec![(10_000, 0.65), (50_000, 0.50), (100_000, 0.25)],
            stock_rate: 0.005,
        };
        // 120_000: 10k*0.65 + 40k*0.50 + 50k*0.25 + 20k*0.25
        let expected = 10_000.0 * 0.65 + 40_000.0 * 0.50 + 50_000.0 * 0.25 + 20_000.0 * 0.25;
        assert!((m.option_cost(10.0, 120_000.0, 100) - expected).abs() < 1e-10);
    }

    #[test]
    fn tiered_stock_cost() {
        let m = CostModel::Tiered {
            tiers: vec![(10_000, 0.65)],
            stock_rate: 0.005,
        };
        assert!((m.stock_cost(150.0, 100.0) - 0.5).abs() < 1e-10);
    }
}
